storage/remote/queue_manager.go

```diff
@@ -466,16 +438,33 @@ func (s *shards) runShard(i int) {
 	}
 }
 
+// sendSamples to the remote storage with backoff for recoverable errors.
 func (s *shards) sendSamples(samples model.Samples) {
-	// Samples are sent to the remote storage on a best-effort basis. If a
-	// sample isn't sent correctly the first time, it's simply dropped on the
-	// floor.
+	backoff := s.qm.cfg.MinBackoff
+	for retries := s.qm.cfg.MaxRetries; retries > 0; retries-- {
+		err := s.sendSamplesOnce(samples)
+		if err == nil {
+			return
+		}
+
+		log.Warnf("error sending %d samples to remote storage: %s", len(samples), err)
+		if _, ok := err.(recorerableError); !ok {
+			return
+		}
+		time.Sleep(backoff)
+		backoff = backoff * 2
+		if backoff > s.qm.cfg.MaxBackoff {
+			backoff = s.qm.cfg.MaxBackoff
+		}
+	}
+}
+
+func (s *shards) sendSamplesOnce(samples model.Samples) error {
 	begin := time.Now()
-	err := s.qm.cfg.Client.Store(samples)
+	err := s.qm.client.Store(samples)
 	duration := time.Since(begin)
 
 	if err != nil {
-		log.Warnf("error sending %d samples to remote storage: %s", len(samples), err)
 		failedSamplesTotal.WithLabelValues(s.qm.queueName).Add(float64(len(samples)))
 	} else {
 		sentSamplesTotal.WithLabelValues(s.qm.queueName).Add(float64(len(samples)))
```

Comment:
That'd make sense to me.

Probably want a separate set of counters for the rpcs.